### PR TITLE
Fix bug in `TensorVariable.__rmatmul__`

### LINKED ISSUE #464: Incorrect shape inferrence for blockwise dot
### Description

It seems some part of the shape inference for blockwise gives incorrect results:

```python
import pytensor
import pytensor.tensor as pt

x = pt.tensor("x", shape=(19, 19))
q = np.random.randn(20, 19)
(q @ x).type
# TensorType(float64, shape=(19, 19))
```

This gives me (19, 19) on main, when it should be (20, 19).

If I wrap `q` into a tensor variable I get the correct result:

```python
import pytensor
import pytensor.tensor as pt

x = pt.tensor("x", shape=(19, 19))
q = pt.as_tensor_variable(np.random.randn(20, 19))
(q @ x).type
# TensorType(float64, shape=(20, 19))
```

### PR DESCRIPTION
**Pull request by ricardoV94**
Introduced in #452 
Closes #464

CC @tomicapretto

### CODE REVIEW

**Comment by jessegrabowski** [comment]
I'm not positive __rmatmul__ is the issue. I put prints into all of __matmul__, __rmatmul__, __dot__, and __rdot__ (I brought these methods back -- I was getting errors in the tests without them?). Only __dot__  was called by the test you flagged as not working.

Digging a bit into the failure case:
```
    X_val = np.arange(2 * 3).reshape((2, 3))
    y_val = np.arange(3)
    res = X_val.dot(y)
    exp_res = dot(X_val, y)
```

exp_res looks like this:

```
[[Mul.0 Mul.0 Mul.0]
 [Mul.0 Mul.0 Mul.0]]
```

The dprint for each cell looks like this (this is cell 0,0)
```
Mul [id A]
 ├─ ExpandDims{axis=0} [id B]
 │  └─ 0 [id C]
 └─ y [id D]
```

Evidently it's broadcasting multiplication of `y` into every cell of `X`, 

```
evaled_res = np.array([[cell.eval({y:y_val}) for cell in row] for row in res])
evaled_res
[[[ 0.  0.  0.]
  [ 0.  1.  2.]
  [ 0.  2.  4.]]

 [[ 0.  3.  6.]
  [ 0.  4.  8.]
  [ 0.  5. 10.]]]

evaled_res.shape
(2, 3, 3)
```

It seems like it's treating X like a batch?

**Comment by ricardoV94** [comment]
To be clear, numpy.array().dot(TensorVariable) doesn't work (and never did AFAICT) and I couldn't see anything we could do without delving into the `__array_func__` stuff. Numpy converts that into a weird rmul by zero, and I couldn't trace that.

Note that `@` is not a dot operation but a matmul operation. Aesara/pytensor had this aspect confused and were using the wrong operator. I flipped the operands of the rmatmul by mistake when I did the fix.

**Comment by ricardoV94** [comment]
To be clear, numpy.array().dot(TensorVariable) doesn't work (and never did AFAICT) and I couldn't see anything we could do without delving into the `__array_func__` stuff. Numpy converts that into a weird rmul by zero, and I couldn't trace that.

Note that `@` is not a dot operation but a matmul operation. Aesara/pytensor had this aspect confused and were using the wrong operator. I flipped the operands of the rmatmul by mistake when I did the fix.

**Comment by ricardoV94** [comment]
About the `__dot__`, `__rdot__`, where are those used? I couldn't find a reference to this being a real dunder method anywhere.

File: tests/tensor/test_variable.py [hunk 1]
```diff
@@ -98,10 +100,15 @@ def test_infix_matmul_method():
     assert equal_computations([res], [exp_res])
 
     X_val = np.arange(2 * 3).reshape((2, 3))
-    res = as_tensor(X_val) @ y
+    res = X_val @ y
```
Comment (ricardoV94) [inline tests/tensor/test_variable.py:103]: This is the test relevant to the reported bug

Comment (jessegrabowski) [reply tests/tensor/test_variable.py:103]: I was running `test_dot_method`, and it failed with an error that `__dot__` was not implemented

**Comment by jessegrabowski** [comment]
I will believe you that np.array @ TensorVariable never worked, but it seems a bit sus. Shouldn't this have come up before if true? PyMC users mix numpy/pytensor objects all the time.

File: tests/tensor/test_variable.py (hunk 1 above)
Comment (ricardoV94) [reply tests/tensor/test_variable.py:103]: Right but if that's the only test that used it's not a reason to have the mysterious dunder method. I couldn't see it being used anywhere else notlr could I see how to trigger it naturally

**Comment by ricardoV94** [comment]
> I will believe you that np.array @ TensorVariable never worked, but it seems a bit sus. Shouldn't this have come up before if true? PyMC users mix numpy/pytensor objects all the time.

@ worked!!! But it used to use pt.dot under the hood and not pt.matmul. I fixed it to use pt.matmul in #452 but messed up the `__rmatmul__` which is called when you do np.array @ TensorVariable

> EDIT: Actually I'm all confused now. @tomicapretto reported that Array @ Tensor (`__matmul__`) doesn't work, but pt.dot(Array, Tensor) does, and you're saying the Array.dot(Tensor) **method** also doesn't work?

`pt.dot` works, what doesn't work is `np.array.dot(TensorVariable)`. It didn't work in Aesara either:

image

**Comment by ricardoV94** [comment]
To recap:

* Calling pytensor functions directly like `pt.dot` or `pt.matmul` always works because it converts the objects to TensorVariables
* Calling @ always works (after this PR), as it will end up in a pt.matmul call
* Calling `np.array(...).dot(pt.tensor(...))` doesn't work and never did AFAICT. This is one example of why I don't like to mix numpy and pytensor objects, because somethings work, like `np.exp(pt.tensor(...))` but not others. This is done via the old [__array_priority__ protocol](https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_priority__) Supposedly the [__array__function__ protocol](https://numpy.org/neps/nep-0018-array-function-protocol.html) is the modern way of overloading numpy expressions when you have custom types, but I never had the time to check how to implement that.

I suspect that would allow us to overload `np.array().dot(...)` correctly.

**Comment by ricardoV94** [comment]
Okay so the `__dot__` methods are used in https://github.com/pymc-devs/pytensor/blob/071eadd8b261d956f5661ca2168c222aa0aea6e1/pytensor/tensor/math.py#L2045-L2066

I reverted their removal

**Review by jessegrabowski — APPROVED** [review]
Thanks for being patient with my questions. This is good to go. I agree the infixes for dot are a strange design choice, but it's a detail for another day.

**Pull request merged**